### ISSUE TITLE
Should do-while loops that look like for-loops be blindly rewritten?

### ISSUE THREAD

**Issue by uxmal** [uxmal/reko#689]
The recent commits will turn the loop
```C
do {
   <loop-body>
   i = i + 1;
} while (i != n);
```
into 
```C
for (; i != n; ++i)
{
    <loop-body>
}
```
which is not correct, because the `do`-loop will always execute the `<loop-body>` at least once, while the `for`-loop's condition may stop it from looping at all. I have a potential fix on my machine that only performs the rewrite if it can be proved that the initial value of `i` causes the loop condition to evaluate to `true`. In practice, the only way Reko can prove this today is if both `i` is initalized to a constant and is compared to a constant.

The regrettable outcome of this is that many of the `for` loops previously generated now revert back to less readable `do` loops. I'm trying to see if there is any way to salvage this, as much of the benefit of the for-loop rewrite is lost otherwise.

One notion is to add a user switch `--force-dowhile-to-for` or the like, which proceeds to convert the `do` loops to `for` loops by disregarding the possibility that the for loop body might not be executed at all.

Please chime in if you have opinions on this topic.

**Comment by ptomin** [comment]
I can't see any reason to rewrite
```
do {
   <loop-body>
   i = i + 1;
} while (i != n);
```
to
```
for (; i != n; ++i)
{
    <loop-body>
}
```
So what the problem? Do you mean Reko prove method is too weak? And so some loops which is actually `for-loops` can not be rewritten.
Are you sure that
>many of the for loops previously generated now revert back to less readable do loops

are _really_ for-loops which were optimised by compiler to `do-while`? If so then it should be possible to prove that condition is `true`

**Comment by ptomin** [comment]
And , I think, it will work properly on @rfalke 's binary from #680

**Comment by uxmal** [comment]
Turning on the termination test on the regression tests removes 106 "not-provable" `do`-loops, leaving 140 `for` loops. After looking at the loops, I see many unrolled machine language loops (Z80, PDP-11, x86) of the type `djnz`, `rep movsb` and the like. These cannot be translated to `for` loops in a straightforward way.

I do see a few 
```
if (a != b) 
{
    do 
    {
        <loop-body>
        a = a + 1;
    } while (a != b);
}
```
which should translate to:
```
while (a != b)
{
    <loop-body>
    a = a + 1;
}
``` 
and those can easily be made to `for`-loops. 

There is a fair number of:
```
if (<end-expr> != 0) 
{
    do 
    {
        <loop-body>
        i = i + 1;
    } while (<end-expr> > i);
}
```
which I believe also can be converted safely with some work. 

I will commit the conservative changes, which will roll back some of the regression tests, then work on addressing the cases outlined above.

**Comment by rfalke** [comment]
Just my 2 cents.

The goal of the transformation is to increase readability. And this depends very much on the expectation the user has. As a user I would want to see:

1. every for loop has all three parts
2. the update part is simple (i+=N (for -10<=N<=10), i*=2)
3. the condition part is simple (i tested against a constant or a variable which isn't change in the loop). So not a function call or complex conditions like "i%10==0"
4. at least one of the initial value or the condition is a constant. So it is either counting from N upwards to var or from var down to N or from N1 to N2.

So basically the decompiler knows how many iterations are there.

In all other cases I would expect a while or do-while loop.

**Comment by uxmal** [comment]
@rfalke: agreed, and Reko does this now. However, @ptomin is questioning the logic behind turning:
```
int i = 0;
do {
    foo();
    i = i + 1;
} where (i < n);
```
into a `for`-loop because we cannot make the assumption that `i < n` is true if `n` is, say, a parameter to the current procedure. It matches the four critieria you give: 
1. the loop here has an initializer, a test, and an update; 
2. the update `i = i + 1` is simple
3. the condition is simple (test against `n` which doesn't change in the loop)
4. and we know that the initial value is a constant.

However, we don't know whether `n` may be zero on entry of the `do`-loop, so while the `do`-form of the loop always executes the body at least once, the `for`-form
```
for (i = 0; i < n; ++i) {
    foo();
}
```
might not execute the body if `n` has the value 0.  @ptomin 's judgment is that performing the transformation without proving that the condition will be evaluated more than once is an error.

Raimar, you reported in (#682) that gated `do`-loops could be simplified; I made a commit last night that does this. That will take care of many `do` whiles, but I'm afraid that the `do`-loop above will have to remain a `do`-loop unless someone can convince me that it's safe to translate it to a `for`.

**Comment by rfalke** [comment]
It looks like a more detailed for loop stress test for decompiler is needed. It would be separate from the Decompiler Maturity Index (DMI). Now only a bit of time is missing ...

Back to the problem. One could transform

```
int i = 0;
do {
    foo(i);
    i = i + 1;
} where (i < n);
```

to

```
foo(0);
for(int i=1; i<n; i++) {
  foo(i);
}
```

But I'm not sure if this is an improvement or not.

**Comment by uxmal** [comment]
I'm not too keen either. I'm going to leave such loops as is, unless I get a lot of complaints about it. Thanks for your input, all. I'm closing this now.
